eclair-core/src/main/scala/fr/acinq/eclair/payment/receive/MultiPartHandler.scala

```diff
@@ -226,32 +270,50 @@ object MultiPartHandler {
-              val paymentMetadata = hex"2a"
-              val invoiceFeatures = if (nodeParams.enableTrampolinePayment) {
-                Features[InvoiceFeature](nodeParams.features.invoiceFeatures().activated + (Features.TrampolinePaymentPrototype -> FeatureSupport.Optional))
+              val featuresTrampolineOpt = if (nodeParams.enableTrampolinePayment) {
+                nodeParams.features.invoiceFeatures().add(Features.TrampolinePaymentPrototype, FeatureSupport.Optional)
               } else {
                 nodeParams.features.invoiceFeatures()
               }
-              val invoice = Bolt11Invoice(
-                nodeParams.chainHash,
-                amount_opt,
-                paymentHash,
-                nodeParams.privateKey,
-                description,
-                nodeParams.channelConf.minFinalExpiryDelta,
-                fallbackAddress_opt,
-                expirySeconds = Some(expirySeconds),
-                extraHops = extraHops,
-                paymentMetadata = Some(paymentMetadata),
-                features = invoiceFeatures
-              )
-              context.log.debug("generated invoice={} from amount={}", invoice.toString, amount_opt)
-              nodeParams.db.payments.addIncomingPayment(invoice, paymentPreimage, paymentType)
-              invoice
+              receivePayment match {
+                case r: ReceiveStandardPayment =>
+                  val expirySeconds = r.expirySeconds_opt.getOrElse(nodeParams.invoiceExpiry.toSeconds)
+                  val paymentMetadata = hex"2a"
+                  val invoice = Bolt11Invoice(
+                    nodeParams.chainHash,
+                    r.amount_opt,
+                    paymentHash,
+                    nodeParams.privateKey,
+                    r.description,
+                    nodeParams.channelConf.minFinalExpiryDelta,
+                    r.fallbackAddress_opt,
+                    expirySeconds = Some(expirySeconds),
+                    extraHops = r.extraHops,
+                    paymentMetadata = Some(paymentMetadata),
+                    features = featuresTrampolineOpt.remove(Features.RouteBlinding)
+                  )
+                  context.log.debug("generated invoice={} from amount={}", invoice.toString, r.amount_opt)
+                  nodeParams.db.payments.addIncomingPayment(invoice, paymentPreimage, r.paymentType)
+                  invoice
+                case r: ReceiveOfferPayment =>
+                  // TODO: get blinded paths from the router instead
+                  val pathId = RouteBlindingEncryptedDataTlv.PathId(randomBytes(32))
+                  val dummyConstraints = RouteBlindingEncryptedDataTlv.PaymentConstraints(CltvExpiry(nodeParams.currentBlockHeight + 144), 1 msat)
+                  val dummyRelay = RouteBlindingEncryptedDataTlv.PaymentRelay(CltvExpiryDelta(0), 0, 0 msat)
+                  val dummyScid = RouteBlindingEncryptedDataTlv.OutgoingChannelId(ShortChannelId(0))
```

Comment:
There can be a collision with a local alias. In https://github.com/ACINQ/eclair/pull/2401, I've used `val toSelf: Alias = Alias(aliasUpperBound)`.
You also need to add some code in `IncomingPaymentPacket.decrypt` to peel one more layer when we get this alias.

Reply:
My goal really wasn't to make workable blinded paths here, I was deferring that to a PR that would implement generating real blinded paths in the router. This PR in itself is incomplete, I was expecting us to integrate this one, your two PRs for sending (#2395) and receiving (#2401, which could add the decryption of dummy hops) and another PR to generate blinded paths before we can actually test this end-to-end, how does that sound?

Reply:
Note that all of this is currently dead code, since `PaymentPacket.decrypt` just rejects blinded final payloads. My plan was that your two PRs (and the final one to create real blinded paths) would build on top of this to get us to feature complete.